File: sdk/integrations/langchain/budgent_langchain.ts
/**
 * Budgent x LangChain.js — a DynamicStructuredTool that lets a LangChain agent
 * pay through Budgent. Non-custodial: the agent only holds a scoped API key, the
 * on-chain program enforces the budget.
 *
 *   npm i budgent-sdk @langchain/core zod
 */
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { BudgentClient, type BudgentConfig } from 'budgent-sdk';

const paySchema = z.object({
  amount: z.number().positive().describe('Amount to pay (in the vault asset units, e.g. USDC).'),
  domain: z.string().optional().describe('Resource domain resolved via the vault registry, e.g. "api.openai.com".'),
  recipient: z.string().optional().describe('Direct recipient pubkey, if not using a domain.'),
  resource: z.string().optional().describe('What is being paid for, e.g. "gpt-4o tokens".'),
  taskId: z.string().optional().describe('Caller task id for attribution / idempotency.'),
});

/**
 * Build the `budgent_pay` tool. The handler returns a concise human/LLM-readable
 * string (status + reason + signature) so the agent can reason about the verdict.
 */
export function budgentPayTool(cfg: BudgentConfig) {
  const client = new BudgentClient(cfg);
  return new DynamicStructuredTool({
    name: 'budgent_pay',
    description:
      'Pay for a resource through Budgent (non-custodial, on-chain enforced). ' +
      'Provide an amount and either a domain or a recipient. Returns the on-chain verdict.',
    schema: paySchema,
    func: async ({ amount, domain, recipient, resource, taskId }) => {
      try {
        const r = await client.pay({ amount, domain, recipient, resource, taskId });
        const parts = [`status=${r.status}`, `reason=${r.reason}`, `amount=${r.amount} ${r.asset}`];
        if (r.signature) parts.push(`signature=${r.signature}`);
        if (r.explorer) parts.push(`explorer=${r.explorer}`);
        return parts.join(' | ');
      } catch (e: any) {
        return `budgent_pay error: ${e?.message ?? String(e)}`;
      }
    },
  });
}

/** Read the agent's own budget snapshot (policy + balance + spend). */
export function budgentBalanceTool(cfg: BudgentConfig) {
  const client = new BudgentClient(cfg);
  return new DynamicStructuredTool({
    name: 'budgent_balance',
    description: 'Get the agent\'s Budgent budget snapshot (policy, balance, spend so far).',
    schema: z.object({}),
    func: async () => {
      try {
        return JSON.stringify(await client.me());
      } catch (e: any) {
        return `budgent_balance error: ${e?.message ?? String(e)}`;
      }
    },
  });
}

File: sdk/integrations/langchain/budgent_langchain.py
"""Budgent x LangChain (Python) — StructuredTools that let a LangChain agent pay
through Budgent. Non-custodial: the agent holds only a scoped API key; the on-chain
program enforces the budget.

    pip install langchain-core pydantic
    # plus budgent.py on the path (the Budgent Python SDK)
"""
from __future__ import annotations

from typing import Optional

from langchain_core.tools import StructuredTool
from pydantic import BaseModel, Field

from budgent import Budgent


class _PayArgs(BaseModel):
    amount: float = Field(..., description="Amount to pay (vault asset units, e.g. USDC).")
    domain: Optional[str] = Field(None, description='Resource domain resolved via the vault registry, e.g. "api.openai.com".')
    recipient: Optional[str] = Field(None, description="Direct recipient pubkey, if not using a domain.")
    resource: Optional[str] = Field(None, description='What is being paid for, e.g. "gpt-4o tokens".')
    task_id: Optional[str] = Field(None, description="Caller task id for attribution / idempotency.")


def _verdict_str(r: dict) -> str:
    parts = [
        f"status={r.get('status')}",
        f"reason={r.get('reason')}",
        f"amount={r.get('amount')} {r.get('asset')}",
    ]
    if r.get("signature"):
        parts.append(f"signature={r['signature']}")
    if r.get("explorer"):
        parts.append(f"explorer={r['explorer']}")
    return " | ".join(parts)


def budgent_pay_tool(base_url: str, key_id: str, hmac_secret: str) -> StructuredTool:
    """Build the `budgent_pay` StructuredTool. Returns a concise verdict string."""
    client = Budgent(base_url=base_url, key_id=key_id, hmac_secret=hmac_secret)

    def _pay(amount: float, domain: Optional[str] = None, recipient: Optional[str] = None,
             resource: Optional[str] = None, task_id: Optional[str] = None) -> str:
        try:
            r = client.pay(
                amount=amount,
                domain=domain,
                recipient=recipient,
                resource=resource or "",
                task_id=task_id or "",
            )
            return _verdict_str(r)
        except Exception as e:  # noqa: BLE001
            return f"budgent_pay error: {e}"

    return StructuredTool.from_function(
        func=_pay,
        name="budgent_pay",
        description=(
            "Pay for a resource through Budgent (non-custodial, on-chain enforced). "
            "Provide an amount and either a domain or a recipient. Returns the on-chain verdict."
        ),
        args_schema=_PayArgs,
    )


def budgent_balance_tool(base_url: str, key_id: str, hmac_secret: str) -> StructuredTool:
    """Build the `budgent_balance` StructuredTool (GET /v1/me)."""
    client = Budgent(base_url=base_url, key_id=key_id, hmac_secret=hmac_secret)

    def _balance() -> str:
        try:
            return str(client.me())
        except Exception as e:  # noqa: BLE001
            return f"budgent_balance error: {e}"

    return StructuredTool.from_function(
        func=_balance,
        name="budgent_balance",
        description="Get the agent's Budgent budget snapshot (policy, balance, spend so far).",
    )
